### ISSUE TITLE
Support for multiple, different LockProviders (i.e. in a modular monolith - each module providing its own locks storage)

### ISSUE THREAD

**Issue by risto-uibo-inbank** [lukas-krecan/ShedLock#2217]
Currently 

- `@SchedulerLock` behind the scenes requires a single, application-wide `LockProvider` bean
- In other words - it's not possible to specify, what locks storage (what kind of lock provider) to use for different scheduled jobs, when applying `@SchedulerLock` annotation on them.
- The annotation processing logic behind the annotation always expects exactly one, single `LockProvider` bean from the Spring context

Issue

- This creates a runtime decoupling problem in modular monoliths. Especially, in projects applying vertical-slice architecture. 
- In this architecture, the application is composed of independent modules - along the logical bounded contexts (BC-s) of the system.
- Each BC  "module" is either a completely separate gradle/maven module, or a distinct root package of the application.
- Each BC module contains the application layer & application config (including Spring beans), core (domain) layer and data providers / adapters (repositories) layer. 
- So each module is as independent as possible - both in terms of its code dependencies as well as runtime dependencies (application context).
- Since currently only a single `LockProvider` bean can (and must) be configured, it must be a shared one between the modules - an application-wide bean. While, each module usually wants to have its own DB schema where it needs to store locks, or use a completely different locking mechanism than other modules, for various reasons. Currently that's not possible.

Solution

- Extend the `@SchedulerLock` annotation to support specifying its `LockProvider`, by a `lockProvider=` qualifier property, which would be the name of the `LockProvider` bean that is to be used here.
- In the annotation processing, the qualified `LockProvider` bean would be requested from Spring Application Context - and the LockManager which executes the scheduled code, is created with this specific LockProvider.

Alternatives
- As in my work, we've come across this need, we have created a sort of a parallel annotation to the `@SchedulerLock` annotation - `@QualifiedSchedulerLock`, which does all the same things, except it also requires specifying the `LockProvider`:

```
@Target({METHOD, ANNOTATION_TYPE})
@Retention(RUNTIME)
public @interface QualifiedSchedulerLock {

  String lockProvider();

  String name();

  String lockAtMostFor() default "";

  String lockAtLeastFor() default "";
}
```

- And to go along with that annotation, a processing aspect, where the main logic snippet is:
```
@Around("@annotation(com.our.company.....schedulerlock.QualifiedSchedulerLock)")
  public Object acquireQualifiedSchedulerLock(ProceedingJoinPoint scheduledMethod) {
    var signature = (MethodSignature) scheduledMethod.getSignature();
    var lockParameters = signature.getMethod().getAnnotation(QualifiedSchedulerLock.class);

    var lockProviderName = lockParameters.lockProvider();
    var lockProvider = (LockProvider) applicationContext.getBean(lockProviderName);

    var lockConfigurationExtractor = configurationExtractorWith(lockParameters);
    var lockManager = createLockManager(lockProvider, lockConfigurationExtractor);

    try {
      lockManager.executeWithLock(running(scheduledMethod));
    } catch (Exception e) {
      log.error("Error occurred while acquiring scheduler lock", e);
    }

    return null;
  }
```

Not the entire code, but the general idea of what it does, should be clear of the snippet.

This enables us to define different LockProvider beans for different BC modules, which for example store locks in their own dedicated DB schemas, in their own lock tables - or if necessary, use a completely other locking mechanism by providing another type of LockProvider.

**Comment by lukas-krecan** [comment]
Hi, thanks for reporting, I will look into that

**Comment by lukas-krecan** [comment]
Hi, would something like [this](https://github.com/lukas-krecan/ShedLock/pull/2222/files#diff-8d711b9a693f2d44dcc7ddcaeef13f1c45ca9d24e6b23c06bb26f24443774a22R57) work for you?

**Comment by risto-uibo-inbank** [comment]
> Hi, would something like [this](https://github.com/lukas-krecan/ShedLock/pull/2222/files#diff-6e1150a36190815527417fe73878fdba7109499bba57d60424c872da8cb0c4d4R76) work for you?

Hi, yes absolutely - this looks great 👍 nice work!

**Comment by lukas-krecan** [comment]
Fixed in 6.0.1
